### DISCUSSION TITLE
Write to Inverter to set Force Charge Times?

### DISCUSSION THREAD

**Discussion by C00K50N** [StealthChesnut/HA-FoxESS-Modbus#54] [Q&A]
If I have understood everything correctly, this integration can only read data from the inverter, not write to it; is that correct?  Does the alternative FoxxESS-HA integration allow writing?  Specifically, can we use HA to specify force charge times?

If so, is it likely to cause a problem having both the Modbus and the other integration installed on the same HA setup?  I'd love to read the data via this Modbus integration, but use the other one to write since this one can't.

And if that's possible, I'd love to set up an automation which reads a) the following day's forecast generation and b) the following day's forecast temperature (which will affect whether the heating is likely to come on) and adjust the force charge times accordingly to charge the batteries by just the right amount.

Should this be possible?

**Comment by StealthChesnut** [comment]
This integration should not conflict with anything else. This is just reading data from the inverter, so if it gets updated from another integration or through the Fox Cloud then this integration isn't taking any conflicting actions.

It's not possible to write / update settings (yet) as I've not found where the data is within the modbus space. Finding the right addresses is a matter of trial and error - Fox won't release any documentation (to me) about it despite it being in the marketing/user manuals as being a feature.

**Comment by nathanmarlor** [comment]
> If I have understood everything correctly, this integration can only read data from the inverter, not write to it; is that correct? Does the alternative FoxxESS-HA integration allow writing? Specifically, can we use HA to specify force charge times?
> 
> If so, is it likely to cause a problem having both the Modbus and the other integration installed on the same HA setup? I'd love to read the data via this Modbus integration, but use the other one to write since this one can't.
> 
> And if that's possible, I'd love to set up an automation which reads a) the following day's forecast generation and b) the following day's forecast temperature (which will affect whether the heating is likely to come on) and adjust the force charge times accordingly to charge the batteries by just the right amount.
> 
> Should this be possible?

@C00K50N regarding your final point, I've just finished a HA integration which does ~90% of what you're asking for -> https://github.com/nathanmarlor/foxess_em

**Reply by C00K50N** [reply]
That's great, thank you.  I'll take a look...

**Reply by nathanmarlor** [reply]
Thanks. It's been doing a great job on my HA for a while, so keen to hear other peoples experiences!

**Reply by spmaid** [reply]
@nathanmarlor so you've managed to write back to the Inverter to force charge? That's great if so!  So can I use your integration along side this one, or is it one or the other?

**Reply by nathanmarlor** [reply]
@spmaid not directly using Modbus but by issuing commands to the Fox Cloud. I'd recommend using this integration alongside mine, they're completely compatible. Once @StealthChesnut has worked out how to set force charge parameters over modbus (;-)) I'll likely be updating my integration to rip out the cloud calls.

**Reply by spmaid** [reply]
Understood, thanks!

**Reply by C00K50N** [reply]
@nathanmarlor just looking at your integration now.  Does it require a Solcast subscription / device?  I'm not familiar with Solcast but was hoping to use the forecast generation figures I already have in HA (from Forecast.Solar I believe).  Will your integration still work?

**Reply by nathanmarlor** [reply]
It does require a Solcast account but you can set one up for free at - https://toolkit.solcast.com.au/live-forecast

I found the Solcast estimates to be much more reliable than the inbuilt integration.

**Reply by C00K50N** [reply]
I have re-installed the Fox Cloud integration to complement the Modbus one.  I've also now got a Solcast account and site set up, and your integration downloaded via HACS @nathanmarlor.  Unfortunately, I'm now stuck I'm afraid... how do I find the Solcast Setup, FoxESS Cloud Setup, etc elements you show screenshots of in the Configuration and Options part of your ReadMe?

(I realise this question should really be part a discussion on the page for your own integration, but forgive me I couldn't find a Discussion section.  I really am just muddling through all of this!)

**Reply by C00K50N** [reply]
Found it!  Sorry, I realised I missed a step - had to Add Integration after downloading via HACS.  So I have all the sensors now at least...  

But... what's next?  Will it now automatically set the force charge times, or is there a further step for me to set that up?  

We usually run other house load on the cheap rate too, in order to try and not cycle the battery more than necessary.  So I've been using both schedules on the Fox app - one to force charge for a short period of time and the other one just to prevent discharge until the end of the cheap period.  Will that be changed automatically now by this integration?

**Reply by nathanmarlor** [reply]
@C00K50N you're right, the integration will now manage the force charge modes for you depending on the solar forecast for the following day. You can see what it's planning on the attributes against the 'Capacity: Charge Needed' sensor - i.e. charge start time, target percentage.

Load shifting appliances to the off-peak period won't upset it, as it calculates house load as the peak period only. You could continue to do this, or if it's preferable (and you have enough battery capacity) move back to running them in the day and the integration will adapt and include the required charge (it averages peak load over 2 days).

**Reply by C00K50N** [reply]
That's brilliant!  Thanks @nathanmarlor.  Yeah, looks like it worked last night, in that it forced a charge right at the end of our cheap period.  Maybe more charge than needed given the good forecast today, but I wonder if it was reacting to something else...  

Is anyone else here seeing the random but significant drop in SoC in the early hours of the morning?  It's happened to me the last 3 nights when I haven't had a force charge scheduled.  Last night from 03:00-05:10 it dropped from 53-26% with no corresponding load or export.  I read of it happening to others on the FaceBook group, maybe some sort of calibration?  But it doesn't add up to me.  That's >4kWh of charge disappeared into the ether!

It seems your integration managed to adapt nonetheless, and it charge up to 44% by 06:30 so we didn't risk being left too low by sunrise.

Great stuff about the load only calculating the peak too.  It seems you've used the same logic I would have had I been able to write anything like this myself!  Thank you.

**Reply by nathanmarlor** [reply]
@C00K50N great to hear it's working for you! You may be seeing more charge than you expect due to the dawn/day buffer. Dawn buffer defaults at 1kW and day buffer at 2kW. 
On a perfect day where the forecast is exactly right and your usage mirrors the previous two days the battery should hit dawn time with 1kW left and should hit the start of off-peak with 2kW left. I put this in as in reality you'll see quite a lot of variation in forecast and usage, and it's better to top-up with cheap off-peak than run out and be forced back onto peak rates :)

**Reply by spmaid** [reply]
> That's brilliant! Thanks @nathanmarlor. Yeah, looks like it worked last night, in that it forced a charge right at the end of our cheap period. Maybe more charge than needed given the good forecast today, but I wonder if it was reacting to something else...
> 
> Is anyone else here seeing the random but significant drop in SoC in the early hours of the morning? It's happened to me the last 3 nights when I haven't had a force charge scheduled. Last night from 03:00-05:10 it dropped from 53-26% with no corresponding load or export. I read of it happening to others on the FaceBook group, maybe some sort of calibration? But it doesn't add up to me. That's >4kWh of charge disappeared into the ether!
> 
> It seems your integration managed to adapt nonetheless, and it charge up to 44% by 06:30 so we didn't risk being left too low by sunrise.
> 
> Great stuff about the load only calculating the peak too. It seems you've used the same logic I would have had I been able to write anything like this myself! Thank you.

Oh, so it's not just me!  I thought I had done something wrong, I regulary lose 25% of the SoC overnight, from about 2am onwards, I have a support email conversation going on for quite a while now with no fix in sight.  I also note that when charging, it will go from say 80% to 99% in a matter of minutes, like the 'missing' SoC comes back.  Its really frustrating as it makes it really difficult to plan if I need to charge the battery overnight.  

I also have another even more annoying issue in that if I get to 19% SoC it starts charging itself from the grid at about 1.5kW! So I have to now set the minimum SoC to about 22% to make sure it doesnt hit 19% otherwise it gets stuck charging, and I have to power down the whole setup to stop it.  That I find even more annoying as I used to run to 10% SoC

**Reply by StealthChesnut** [reply]
@spmaid I have found with my batteries I had it set to 15% min SoC, but overnight it would lose enough to start force charging as you describe. I've upped mine to 20% min SoC, but now it is back to doing the same thing again. It'll discharge to 20%, then lose another 5-6% over the course of a few hours and then start to force charge again.

The only thing I could think of to stop this would be to update the system to track the min SoC gradually from a "preferred value" (e.g. 20%) to an absolute minimum (e.g. 10%) as the actual SoC passes through the preferred min SoC. It seems that once you get 5% below the set value it'll start charging. If the set value was updated to be +2% (so when SoC reaches 16%, set the min SoC to be 17%) it wouldn't allow for discharging the battery but would stop it from charging from mains.

However, if it reaches the absolute min SoC and solar forecast is bad then it's probably worth resetting back to a higher value and allowing the charge again.

**Reply by C00K50N** [reply]
Definitely not just you @spmaid.  Here's one thread on FaceBook where a number of people have the same issue: https://www.facebook.com/groups/foxessownersgroup/permalink/1657902311258611/

**Reply by C00K50N** [reply]
@nathanmarlor are you able to set up a discussion page on your integration?  I have another query or two and feel bad for constantly asking about it on here.

**Reply by nathanmarlor** [reply]
@C00K50N done!

**Comment by Daibutt** [comment]
I'd agree, Solcast for my PV was a lot more accurate than Forecast.Solar, I've been using the Rooftops solar bit https://toolkit.solcast.com.au/rooftop-sites, I guess that's the same as @nathanmarlor linked to above (still free for max 50 API calls/day).
